sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnarBatch.java

```diff
@@ -82,25 +91,53 @@ public void close() {
    * Adapter class to interop with existing components that expect internal row. A lot of
    * performance is lost with this translation.
    */
-  public final class Row extends InternalRow {
+  public static final class Row extends InternalRow {
     private int rowId;
+    private final ColumnarBatch parent;
+    private final int fixedLenRowSize;
+
+    private Row(ColumnarBatch parent) {
+      this.parent = parent;
+      this.fixedLenRowSize = UnsafeRow.calculateFixedPortionByteSize(parent.numCols());
+    }
 
     /**
      * Marks this row as being filtered out. This means a subsequent iteration over the rows
      * in this batch will not include this row.
      */
     public final void markFiltered() {
-      ColumnarBatch.this.markFiltered(rowId);
+      parent.markFiltered(rowId);
     }
 
     @Override
     public final int numFields() {
-      return ColumnarBatch.this.numCols();
+      return parent.numCols();
     }
 
     @Override
+    /**
+     * Revisit this. This is expensive.
+     */
     public final InternalRow copy() {
-      throw new NotImplementedException();
+      UnsafeRow row = new UnsafeRow(parent.numCols());
```

Comment:
Is that valid for copy? Copy seems like it can't have those semanatics. I think we need to remove copy() from the row interface internally. Or at least change it to take to something like:

void copy(UnsafeRowWriter writer). We should be able to directly write this to the destination.